Lib/cProfile.py

```diff
@@ -179,10 +178,21 @@ def main():
                 '__name__': spec.name,
                 '__package__': None,
                 '__cached__': None,
+                '__builtins__': __builtins__,
             }
+        # cmd has to run in __main__ namespace (or imports from __main__ will
+        # break). Clear __main__ and replace with the globals provided.
+        import __main__
+        # Save a reference to the current __main__ namespace so that we can
+        # restore it after cmd completes.
+        original_main = __main__.__dict__.copy()
+        __main__.__dict__.update(globs)
+
         try:
-            runctx(code, globs, None, options.outfile, options.sort)
+            runctx(code, __main__.__dict__, None, options.outfile, options.sort)
```

Comment:
Why do you need to restore it? You are exiting the program anyway. Also this is not ideal either. This will include all the global variables to the script that is being profiled. We want `print(locals())` to be basically the same with or without the profiler.

Reply:
The issue is to juggle the global variables needed by the call to (and implementation of) runctx. It would be a lot easier if I could split cprofile into a module where __main__ only has the main function. Is that something I can do?
if not, it's still possible, just trickier.

Reply:
No you can't split it into its own module :( that's too much a change. I think the correct way to go is to make the full `runctx` path independent of any global variables.

Reply:
Found a much cleaner fix - just ensure that the "main" function isn't run in the `__main__` namespace.

Reply:
Hmm, I'm not sure if this is too hacky. I did not find such pattern in other code. It looks like an acceptable solution but I don't know if there will be implications. I want to ask @vstinner about this as he probably knows a lot of interesting usages.

Reply:
Is there a formal way for me to request a review from them?

Reply:
Isn't this what pdb already does implicitly?

Reply:
https://github.com/python/cpython/blob/de9deb7ca7120fbb5cbbb53044ce91087065e723/Lib/pdb.py#L2449

Ah, it's actually explicitly documented

Reply:
I think the current solution looks much less hacky.